src/Compilers/Core/MSBuildTask/ManagedCompiler.cs

```diff
@@ -287,7 +287,7 @@ protected override Encoding StandardOutputEncoding
 
         protected override int ExecuteTool(string pathToTool, string responseFileCommands, string commandLineCommands)
         {
-            if (!UseSharedCompilation)
+            if (!UseSharedCompilation || this.ToolPath != null )
```

Comment:
pathToTool is fully resolved using a bunch of defaults and heuristics baked into the superclasses. It should usually resolve to the local MSBuild folder when ToolPath isn't set, provided you're not doing anything super strange.

ToolPath is meant to be the entry point for defining your own location for the compiler executable. We want to respect people setting CscToolPath and not use the client in the task.